File: mqtt_panel/web/widget/iframe.js
$(function() {
    $('.widget-iframe').on('update', function(event, blob) {
        $(this).data('mtime', blob.mtime);
        let src_change = $(this).find('iframe').attr('src') != blob.src;
        if (src_change) {
            $(this).find('iframe').attr('src', blob.src)
        }

        $(this).trigger('enable');

        if (this.interval) {
            if (src_change == false) {
                return;
            }
            clearInterval(this.interval);
            this.interval = null;
        }
        let iframe = $(this).find('iframe');
        let refresh = parseInt($(iframe).data('refresh'));

        if (refresh) {
            let This = this;
            this.interval = setInterval(function() {
                $(iframe).attr('src', function(_, val) {
                    return val;
                });
                let now = Date.now()/1000;
                $(This).data('mtime', now);
                widget_update_last_update(This, now);
            }, refresh * 1000);
        }
    });

    $('.widget-iframe').on('disable', function(event, blob) {
        if (this.interval) {
            clearInterval(this.interval);
            this.interval = null;
        }
        let iframe = $(this).find('iframe');
        $(iframe).attr('src', '');
    });
});

File: mqtt_panel/web/widget/iframe.py
import logging
from mqtt_panel.web.widget.widget import Widget


class Iframe(Widget):
    widget_type = 'iframe'

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._src = self._c['attr'].get('src', '')

    def open(self):
        topic = self._c.get('subscribe', None)
        if topic:
            self._mqtt.subscribe(topic, self._on_mqtt)

    def _on_mqtt(self, payload, _timestamp):
        logging.info("{%s} Rx MQTT: %s", self.id, payload)
        if payload == self._src:
            return
        self._src = payload
        self._updated_now()
        self._update_clients()

    def _blob(self):
        return {
            'src': self._src
        }

    def _html(self, fh):
        refresh = self._c.get('refresh', '')
        if refresh:
            refresh = f' data-refresh="{refresh}"'
        self._write_render(fh, f'<iframe src=""{refresh}')
        for key, value in self._c.get('attr', []).items():
            if key == 'src':
                continue
            if self._c['attr'].get(key, None):
                fh.write(f' {key}="{value}"')
            else:
                fh.write(f' {key}')
        fh.write(' frameborder="0"></iframe>\n')
